### PR TITLE
python3Packages.llmx: init at 0.0.21a0

### LINKED ISSUE #289897: Package request: python3.llmx
**Project description**
 An API for Chat Fine-Tuned Large Language Models (llm) 

A simple python package that provides a unified interface to several LLM providers of chat fine-tuned models [OpenAI, AzureOpenAI, PaLM, Cohere and local HuggingFace Models].

**Metadata**

* homepage URL: https://pypi.org/project/llmx/
* source URL: https://github.com/victordibia/llmx
* license: mit
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Moraxyc**
close #289897

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/llmx/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchPypi,
+  setuptools-scm,
+  setuptools,
+  python3Packages,
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:7]:
```suggestion
```

File: pkgs/development/python-modules/llmx/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  pname = "llmx";
+  version = "0.0.21a0";
+  pyproject = true;
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-OEo6wIaDTktzAsP0rOmhxjFSHygTR/EpcRI6AXsu+6M=";
+  };
+
+  disabled = python3Packages.pythonOlder "3.9";
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:20]:
```suggestion
  disabled = pythonOlder "3.9";
```

File: pkgs/development/python-modules/llmx/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  dependencies = with python3Packages; [
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:22]:
```suggestion
  dependencies = [
```

File: pkgs/development/python-modules/llmx/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    tiktoken
+    diskcache
+    cohere
+    google-auth
+    typer
+    pyyaml
+    setuptools
+  ];
+
+  build-system = [ setuptools-scm ];
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:34]: It's preferred to keep the inputs in logical order as they are used. You are free to do it the way you like but for reviewer is more convenient if the structure is similar to what is present.

File: pkgs/development/python-modules/llmx/default.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  meta = with lib; {
+    description = "LLMX: A library for LLM Text Generation";
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:37]:
```suggestion
    description = "Library for LLM Text Generation";
```

see https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

File: pkgs/development/python-modules/llmx/default.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    openai
+    tiktoken
+    diskcache
+    cohere
+    google-auth
+    typer
+    pyyaml
+    setuptools
+  ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:33]: There are [optional dependencies](https://github.com/victordibia/llmx/blob/main/pyproject.toml#L31C1-L31C114).

**Review by fabaff — COMMENTED** [review]
There are tests present. At first glance, using `pytestCheckHook` should execute them.

Please add `pythonImportsCheck`.

**Comment by Moraxyc** [comment]
> There are tests present. At first glance, using `pytestCheckHook` should execute them.
> 
> Please add `pythonImportsCheck`.

Tests try to access OpenAI, google with API from env.

File: pkgs/development/python-modules/llmx/default.nix [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      fastapi
+      uvicorn
+    ]
+    ++ lib.optionals withTransformers [
+      transformers
+      accelerate
+    ];
+
+  # Tests of llmx try to access openai, google, etc.
+  # nativeCheckInputs = [pytestCheckHook];
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:60]: This makes sure that no tests are executed.

```suggestion
  doCheck = false;
```

File: pkgs/development/python-modules/llmx/default.nix [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      setuptools
+    ]
+    ++ lib.optionals withWeb [
+      fastapi
+      uvicorn
+    ]
+    ++ lib.optionals withTransformers [
+      transformers
+      accelerate
+    ];
```
Comment (fabaff) [inline pkgs/development/python-modules/llmx/default.nix:57]: Remove `withWeb` and `withTransformers` and use `optional-dependencies` please. With `optional-dependencies` will consumers be able to pull what they need `++ llmx.optional-dependencies.web` or `++ llmx.optional-dependencies.transformers`.

```suggestion
    ];

  passthru.optional-dependencies = {
    web = [
       fastapi
       uvicorn
    ];
    transformers = [
      accelerate
      transformers
    ] ++ transformers.optional-dependencies.torch;
  };
```

File: pkgs/development/python-modules/llmx/default.nix [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  dependencies = [
+    pydantic
+    openai
+    tiktoken
+    diskcache
+    cohere
+    google-auth
+    typer
+    pyyaml
+    setuptools
```
Comment (natsukium) [inline pkgs/development/python-modules/llmx/default.nix:45]: `setuptools` should go to `build-system`.

**Comment by natsukium** [comment]
@ofborg build python311Packages.llmx python312Packages.llmx

**Review by natsukium — APPROVED** [review]
LGTM, thanks.

**Pull request merged**